src/main/store.ts

```diff
@@ -78,50 +76,84 @@ export class Store<T> {
    * from disk. If the file does not exist, the initial value provided to the
    * constructor is used.
    */
-  public get(): T {
-    if (this._data === undefined) {
-      try {
-        this._data = existsSync(this._path)
-          ? (JSON.parse(readFileSync(this._path, 'utf8'), dateReviver) as T)
-          : this._initial;
-      } catch (e) {
-        this._data = this._initial;
+  public async get(): Promise<T> {
+    return await this._lock.runExclusive(async () => {
+      if (this._data === undefined) {
+        try {
+          this._data = JSON.parse(await readFileAsync(this._path, 'utf8'), dateReviver) as T;
+        } catch (e) {
+          this._data = this._initial;
+        }
       }
-    }
 
-    return this._data;
+      return this._data;
+    });
+  }
+
+  /**
+   * Updates data by passing it through the given function.
+   * @param fn A function receiving the current data and returning new one.
+   */
+  public async update(fn: (current: T) => T): Promise<void> {
+    await this.set(fn(await this.get()));
   }
 
   /** Returns store to its initial state */
-  public clear(): void {
-    this.set(this._initial);
+  public async clear(): Promise<void> {
+    await this.set(this._initial);
   }
 
   /** Gets the Date that the file was last modified */
-  public getModifiedDate(): Date | undefined {
+  public async getModifiedDate(): Promise<Date | undefined> {
     try {
-      return statSync(this._path)?.mtime;
+      return (await statAsync(this._path))?.mtime;
     } catch (_) {
       return undefined;
     }
   }
+}
 
-  /** Serializes the current data into the JSON file. */
-  private _flush(): void {
-    try {
-      if (this._data === undefined) {
-        unlinkSync(this._path);
-      } else {
-        mkdirpSync(dirname(this._path));
-        writeFileSync(this._path, JSON.stringify(this._data));
-      }
-    } catch (e) {
-      logger.warn('Failed to flush store', e);
-      // This usually fails due to anti virus scanners, issues in the file
-      // system, or problems with network drives. We cannot fix or handle this
-      // issue and must resume gracefully. Thus, we have to ignore this error.
-    } finally {
-      this._flushing = false;
+/**
+ * Extends Store to throttle writes.
+ */
+export class ThrottledStore<T> extends Store<T> {
+  /** The minimum time between writes */
+  private readonly _throttleTime?: number;
```

Comment:
Essentially, we are using the timer to dedupe batched updates right? Can we just call this `DedupedStore` then?

Reply:
Deduping is "the process of removing identical entries" which is not really what's happening here. 

We are throttling or buffering flushes to disk. 

How about `BufferedWriteStore`?

Reply:
Yes you're right - and `BufferedWriteStore` maybe fits better here. 

I'm just thinking if there's a way we can avoid the 500ms and just have a queue + event loop take care of this for us, having the throttle feels like kind of a code smell to me.

Reply:
> There's no point flushing values to disk if you already know they'll be overwritten

Great point, this convinced me. Let's do this for now and improve later if we get feedback.